### PR TITLE
Issue/6999 cad performance

### LINKED ISSUE #6999: Improve performance of cross agent depdency handling
# Problem 
We are observing scaling problems with CAD's. 

The current mechanism is as follows:
- when a agent starts to deploy, it find all requires that go to another agent
- for those, it constructs a RemoteResourceAction that will 
   - ask the server for the last know state for this resource
   - if that is deployed available or deploying, it indicates it ready
   - otherwise it waits for an event from the server 

When we have many resources with many CAD's (a few thousand) this 'ask the server step' overwhelms the agent max session limit. For this, we added a rate limiter (3 at a time), but now it is just very slow as each call takes up to 50ms (and DB connection pool contention adds to this)

# Solution

We may want to resolve these dependencies in batch. 
Luck would have it that the code on the server that produces the resource for the agent already has to pull in all CAD's to the server. So we can just add this extra info to the return value, at little cost. If does however make all of this a bit more complicated,

### PR DESCRIPTION
**Pull request by wouterdb**
# Description

Attempt to improve CAD performance by batching 

closes #6999 

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [ ] No (preventable) type errors (check using make mypy or make mypy-diff)

```
src/inmanta/protocol/methods_v2.py:1522: error: Missing return statement  [empty-body]
src/inmanta/data/__init__.py:4639: error: Argument 1 to "ResourceIdStr" has incompatible type "object"; expected "str"  [arg-type]
src/inmanta/data/__init__.py:5724: error: Enum index should be a string (actual index type "object")  [misc]
src/inmanta/server/services/resourceservice.py:1245: error: Incompatible return value type (got "dict[ResourceIdStr, ResourceState]", expected "dict[str, ResourceState]")  [return-value]

``` 
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )
- [ ] If this PR fixes a race condition in the test suite, also push the fix to the relevant stable branche(s) (see [test-fixes](https://internal.inmanta.com/development/core/tasks/build-master.html#test-fixes) for more info)

### CODE REVIEW

File: src/inmanta/agent/agent.py [hunk 1]
```diff
@@ -138,12 +135,16 @@ def __init__(self, scheduler: "ResourceScheduler", gid: uuid.UUID, reason: str)
...
+    # resourceid -> attribute -> {current: , desired:}
+    changes: dict[ResourceVersionIdStr, dict[str, AttributeStateChange]]
+
     def __init__(self, scheduler: "ResourceScheduler", resource: Resource, gid: uuid.UUID, reason: str) -> None:
         """
         :param gid: A unique identifier to identify a deploy. This is local to this agent.
         """
         super().__init__(scheduler, resource.id, gid, reason)
         self.resource: Resource = resource
+        self.change: Optional[const.Change] = const.Change.nochange
```
Comment (wouterdb) [inline src/inmanta/agent/agent.py:147]: Moved up from baseclass, not needed there

File: src/inmanta/server/services/resourceservice.py [hunk 2]
```diff
@@ -1231,6 +1231,19 @@ async def resource_list(
 
         # TODO: optimize for no orphans
 
+    @handle(methods_v2.resources_status, env="tid")
+    async def resources_status(
+        self,
+        env: data.Environment,
+        model_version: int,
+        rids: list[ResourceIdStr],
+    ) -> dict[str, ResourceState]:
```
Comment (Hugo-Inmanta) [inline src/inmanta/server/services/resourceservice.py:1240]:
```suggestion
    ) -> dict[ResourceIdStr, ResourceState]:
```

File: src/inmanta/agent/agent.py [hunk 3]
```diff
@@ -599,6 +561,25 @@ async def mark_deployment_as_finished(self, resource_actions: Iterable[ResourceA
...
+                model_version=version,
+                rids=list(cads.keys()),
+            )
+            if result.code != 200 or result.result is None:
+                LOGGER.error("Failed to get the status for remote resources %s (%s)", list(cads.keys()), result.result)
+                return
+
+            for rid_raw, status_raw in result.result["data"].items():
+                status = const.ResourceState[status_raw]
+                rra = cads.get(rid_raw, None)
```
Comment (Hugo-Inmanta) [inline src/inmanta/agent/agent.py:578]: Is there a specific scenario where a resource id would be returned by `resources_status` but not be present in the `cads` dict ?

**Review by Hugo-Inmanta — CHANGES_REQUESTED** [review]

File: src/inmanta/server/services/resourceservice.py (hunk 2 above)
Comment (wouterdb) [reply src/inmanta/server/services/resourceservice.py:1240]: that produces a type error on the @handle annotation, so I had to choose

File: src/inmanta/agent/agent.py (hunk 3 above)
Comment (wouterdb) [reply src/inmanta/agent/agent.py:578]: mhh, right. That shouldn't happen (the other way around can happen)

**Review by Hugo-Inmanta — APPROVED** [review]

File: src/inmanta/data/__init__.py [hunk 4]
```diff
@@ -4621,6 +4621,22 @@ def convert_or_ignore(rvid):
...
+        cls,
+        env: uuid.UUID,
+        model_version: int,
+        rids: list[ResourceIdStr],
+    ) -> dict[ResourceIdStr, ResourceState]:
+        query = (
+            "SELECT r.resource_id, r.status FROM resource r"
+            " WHERE r.environment=$1"
+            " AND r.model=$2"
+            " AND r.resource_id = ANY($3);"
```
Comment (arnaudsjs) [inline src/inmanta/data/__init__.py:4635]:
```suggestion
            """
            SELECT r.resource_id, r.status
            FROM resource r
            WHERE r.environment=$1
                AND r.model=$2
                AND r.resource_id = ANY($3);
            """
```

File: src/inmanta/data/__init__.py [hunk 5]
```diff
@@ -4621,6 +4621,22 @@ def convert_or_ignore(rvid):
...
         return out
 
+    @classmethod
+    async def get_status_for(
+        cls,
+        env: uuid.UUID,
+        model_version: int,
+        rids: list[ResourceIdStr],
+    ) -> dict[ResourceIdStr, ResourceState]:
+        query = (
```
Comment (arnaudsjs) [inline src/inmanta/data/__init__.py:4631]:
```suggestion
    ) -> dict[ResourceIdStr, ResourceState]:
        if not rids:
            return {}
        query = (
```

File: src/inmanta/agent/agent.py [hunk 6]
```diff
@@ -888,6 +840,8 @@ def cancel(self) -> None:
         """
         for ra in self.generation.values():
             ra.cancel()
+        if self.cad_resolver is not None:
+            self.cad_resolver.cancel()
```
Comment (arnaudsjs) [inline src/inmanta/agent/agent.py:844]:
```suggestion
        if self.cad_resolver is not None:
            self.cad_resolver.cancel()
            self.cad_resolver = None
```

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

**Review by arnaudsjs — APPROVED** [review]

**Comment by wouterdb** [comment]
take over by  #7315

**Pull request closed without merging**